homeassistant/components/trafikverket_ferry/sensor.py

```diff
@@ -0,0 +1,166 @@
+"""Train information for departures and delays, provided by Trafikverket."""
+from __future__ import annotations
+
+from datetime import date, datetime, time, timedelta
+import logging
+
+from pytrafikverket import TrafikverketFerry
+from pytrafikverket.trafikverket_ferry import FerryStop
+
+from homeassistant.components.sensor import SensorDeviceClass, SensorEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_API_KEY, CONF_NAME, CONF_WEEKDAY, WEEKDAYS
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.device_registry import DeviceEntryType
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util.dt import UTC, as_utc, parse_time
+
+from .const import CONF_FROM, CONF_TIME, CONF_TO, DOMAIN
+from .util import create_unique_id
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_FROM = "from_harbour"
+ATTR_TO = "to_harbour"
+ATTR_MODIFIED_TIME = "modified_time"
+ATTR_OTHER_INFO = "other_info"
+
+ICON = "mdi:ferry"
+SCAN_INTERVAL = timedelta(minutes=5)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Trafikverket sensor entry."""
+
+    httpsession = async_get_clientsession(hass)
+    ferry_api = TrafikverketFerry(httpsession, entry.data[CONF_API_KEY])
+
+    try:
+        await ferry_api.async_get_next_ferry_stop(entry.data[CONF_FROM])
+    except ValueError as error:
+        if "Invalid authentication" in error.args[0]:
+            raise ConfigEntryAuthFailed from error
+        raise ConfigEntryNotReady(
+            f"Problem when trying station {entry.data[CONF_FROM]} to {entry.data[CONF_TO]}. Error: {error} "
+        ) from error
+
+    ferry_time = parse_time(entry.data[CONF_TIME])
```

Comment:
I don't think `ConfigEntryNotReady`/`ConfigEntryAuthFailed` will work from here.  It needs to happen in `__init__.py` so it can raise before integration setup is completed

Reply:
I believe you're right so it's now removed and instead given also the below comments I implemented a coordinator instead.